serving/src/main/java/feast/serving/service/OnlineServingService.java

```diff
@@ -0,0 +1,162 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package feast.serving.service;
+
+import static feast.serving.util.Metrics.requestCount;
+import static feast.serving.util.Metrics.staleKeyCount;
+import static feast.serving.util.RefUtil.generateFeatureStringRef;
+
+import com.google.common.collect.Maps;
+import com.google.protobuf.Duration;
+import feast.serving.ServingAPIProto.*;
+import feast.serving.ServingAPIProto.GetOnlineFeaturesRequest.EntityRow;
+import feast.serving.ServingAPIProto.GetOnlineFeaturesResponse.FieldValues;
+import feast.serving.specs.CachedSpecService;
+import feast.serving.util.RefUtil;
+import feast.storage.api.retrieval.FeatureSetRequest;
+import feast.storage.api.retrieval.OnlineRetriever;
+import feast.types.FeatureRowProto.FeatureRow;
+import feast.types.ValueProto.Value;
+import io.grpc.Status;
+import io.opentracing.Scope;
+import io.opentracing.Tracer;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+
+public class OnlineServingService implements ServingService {
+
+  private static final Logger log = org.slf4j.LoggerFactory.getLogger(OnlineServingService.class);
+  private final CachedSpecService specService;
+  private final Tracer tracer;
+  private final OnlineRetriever retriever;
+
+  public OnlineServingService(
+      OnlineRetriever retriever, CachedSpecService specService, Tracer tracer) {
+    this.retriever = retriever;
+    this.specService = specService;
+    this.tracer = tracer;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public GetFeastServingInfoResponse getFeastServingInfo(
+      GetFeastServingInfoRequest getFeastServingInfoRequest) {
+    return GetFeastServingInfoResponse.newBuilder()
+        .setType(FeastServingType.FEAST_SERVING_TYPE_ONLINE)
+        .build();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public GetOnlineFeaturesResponse getOnlineFeatures(GetOnlineFeaturesRequest request) {
+    try (Scope scope = tracer.buildSpan("Redis-getOnlineFeatures").startActive(true)) {
+      GetOnlineFeaturesResponse.Builder getOnlineFeaturesResponseBuilder =
+          GetOnlineFeaturesResponse.newBuilder();
+      List<FeatureSetRequest> featureSetRequests =
+          specService.getFeatureSets(request.getFeaturesList());
+      List<EntityRow> entityRows = request.getEntityRowsList();
+      Map<EntityRow, Map<String, Value>> featureValuesMap =
+          entityRows.stream()
+              .collect(Collectors.toMap(row -> row, row -> Maps.newHashMap(row.getFieldsMap())));
+
+      List<List<FeatureRow>> featureRows =
+          retriever.getOnlineFeatures(entityRows, featureSetRequests);
+
+      for (var fsIdx = 0; fsIdx < featureRows.size(); fsIdx++) {
```

Comment:
I think this block of code would benefit from comments that describe what is happening.

Reply:
@zhilingc You added docs that can resolve this comment, but could we now take it a step further and factor some parts out to helpers, or domain methods on classes involved?

Forgive the laundry list, I'm going to make suggestions before I start a line-by-line review that will take me awhile to finish:

- Pull the metrics out to helpers, they eat up a lot of lines especially in deeply-nested code and metrics always add noise distracting readers from the business logic. Style nit while we're on this: I think the static imports are less than helpful, on first pass I was looking for a local for `staleKeyCount`, if the usage site was `Metrics.staleKeyCount` I would not have needed to.
- Scan for variable extractions, there are a few things repeated, e.g. `featureSetRequest.getSpec().getProject()`, `featureSetRequest.getSpec().getMaxAge()`—nicely named locals will reduce the reading fatigue.
- Minor, but the part assigning `Map<String, FeatureReference> featureNames` could perhaps reasonably be an instance method on `FeatureSetRequest`, like `refsByName` or something (also a suggestion for the local identifier name here, `featureNames` sounds like a sequence to me rather than a map).
- `isStale` (which could currently be marked `static` btw) could also _maybe_ be a method on `FeatureSetRequest`, somehow… I don't like that as-is, especially because we've moved it to storage-api in this PR and `EntityRow` is a nested message of `GetOnlineFeaturesRequest`—that does not belong at that layer. Maybe we're missing an abstraction here.

   `isStale` is behavior that should be unit tested. It could be, as a static method of `OnlineServingService`, but it feels wrong: there's nothing binding the arguments as belonging together in an aggregate (in DDD sense of the word). In other words, it's easy for a caller to make a mistake in using it. How could we improve that? I think this belongs in a perhaps-missing domain service layer (again in DDD sense, this `OnlineServingService` is pretty much an MVC controller currently, and that's not where I want to test or rely on core business logic behavior of Feast).

Reply:
Note that what `isStale` really needs is the feature set spec in question, the `FeatureSetRequest` is incidental. That's a hint at the boundaries of a domain service layer, I'd say.

Somewhat random aside, `FeatureSetReference` is currently defined in `feast.serving.ServingAPIProto`—I think it probably belongs in `feast.types`. It was added for v0.4, unfortunate that moving it is probably a breaking change. At least Maven module-wise it's all one dependency on `datatypes-java`, it'll look out of place as an import in layers far removed from serving but otherwise little consequence.

Reply:
`isStale` is definitely a bit of a conundrum here, I don't think it belongs to the `FeatureSetRequest` because I dont think its the responsibility of the request info object (which is for the most part a value object) to contain that sort of logic. 

Regarding `FeatureSetReference`, The recent PR #548 adds this as a datatype ([see here](https://github.com/gojek/feast/blob/1dd0f503a56ce4b55706317d5d4acde227305e91/protos/feast/core/FeatureSetReference.proto)). Once that PR has been merged in I will refactor this code to use it.